server/start.go

```diff
@@ -259,182 +264,99 @@ func startStandAlone(svrCtx *Context, appCreator types.AppCreator) error {
+
+			app.RegisterTxService(clientCtx)
+			app.RegisterTendermintService(clientCtx)
+			app.RegisterNodeService(clientCtx, svrCfg)
 		}
 	}
 
-	// Add the tx service to the gRPC router. We only need to register this
-	// service if API or gRPC is enabled, and avoid doing so in the general
-	// case, because it spawns a new local CometBFT RPC client.
-	if (config.API.Enable || config.GRPC.Enable) && tmNode != nil {
-		// Re-assign for making the client available below do not use := to avoid
-		// shadowing the clientCtx variable.
-		clientCtx = clientCtx.WithClient(local.New(tmNode))
-
-		app.RegisterTxService(clientCtx)
-		app.RegisterTendermintService(clientCtx)
-		app.RegisterNodeService(clientCtx, config)
-	}
-
-	metrics, err := startTelemetry(config)
+	metrics, err := startTelemetry(svrCfg)
 	if err != nil {
 		return err
 	}
 
 	emitServerInfoMetrics()
 
-	var (
-		apiSrv  *api.Server
-		grpcSrv *grpc.Server
-	)
-
 	ctx, cancelFn := context.WithCancel(context.Background())
 	g, ctx := errgroup.WithContext(ctx)
 
 	// listen for quit signals so the calling parent process can gracefully exit
 	ListenForQuitSignals(cancelFn, svrCtx.Logger)
 
-	if config.GRPC.Enable {
-		_, port, err := net.SplitHostPort(config.GRPC.Address)
-		if err != nil {
-			return err
-		}
-
-		maxSendMsgSize := config.GRPC.MaxSendMsgSize
-		if maxSendMsgSize == 0 {
-			maxSendMsgSize = serverconfig.DefaultGRPCMaxSendMsgSize
-		}
-
-		maxRecvMsgSize := config.GRPC.MaxRecvMsgSize
-		if maxRecvMsgSize == 0 {
-			maxRecvMsgSize = serverconfig.DefaultGRPCMaxRecvMsgSize
-		}
-
-		grpcAddress := fmt.Sprintf("127.0.0.1:%s", port)
-
-		// if gRPC is enabled, configure gRPC client for gRPC gateway
-		grpcClient, err := grpc.Dial(
-			grpcAddress,
-			grpc.WithTransportCredentials(insecure.NewCredentials()),
-			grpc.WithDefaultCallOptions(
-				grpc.ForceCodec(codec.NewProtoCodec(clientCtx.InterfaceRegistry).GRPCCodec()),
-				grpc.MaxCallRecvMsgSize(maxRecvMsgSize),
-				grpc.MaxCallSendMsgSize(maxSendMsgSize),
-			),
-		)
-		if err != nil {
-			return err
-		}
-
-		clientCtx = clientCtx.WithGRPCClient(grpcClient)
-		svrCtx.Logger.Debug("gRPC client assigned to client context", "target", grpcAddress)
-
-		grpcSrv, err = servergrpc.NewGRPCServer(clientCtx, app, config.GRPC)
-		if err != nil {
-			return err
-		}
-
-		// Start the gRPC server in a goroutine. Note, the provided ctx will ensure
-		// that the server is gracefully shut down.
-		g.Go(func() error {
-			return servergrpc.StartGRPCServer(ctx, svrCtx.Logger.With("module", "grpc-server"), config.GRPC, grpcSrv)
-		})
+	grpcSrv, clientCtx, err := startGrpcServer(ctx, g, svrCfg.GRPC, clientCtx, svrCtx, app)
+	if err != nil {
+		return err
 	}
 
-	if config.API.Enable {
+	if svrCfg.API.Enable {
+		// TODO: Why do we reload and unmarshal the entire genesis doc in order to get the chain ID.
+		// surely theres a better way. This is likely a serious node start time overhead.
+		genDocProvider := getGenDocProvider(cmtCfg)
```

Comment:
Fixing this (in another PR) would be really helpful!

Maybe should become a follow-up issue?